File: app/src/main/java/loshica/quiz/QuestionActivity.java
package loshica.quiz;

import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager2.widget.ViewPager2;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import java.text.MessageFormat;
import java.util.Objects;

import loshica.quiz.viewModel.Coordinator;
import loshica.quiz.view.FinishFragment;
import loshica.quiz.view.MyPageTransformer;
import loshica.quiz.view.QuestionAdapter;
import loshica.quiz.view.QuestionFragment;
import loshica.vendor.LOSSettingsActivity;
import loshica.vendor.LOSTheme;

public class QuestionActivity extends AppCompatActivity implements
    QuestionFragment.QuestionFragmentListener,
    FinishFragment.FinishFragmentListener {

    int theme;

    ViewPager2 qp;
    QuestionAdapter qa;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // LOSTheme
        theme = new LOSTheme(this).current;
        setTheme(theme);
        //

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_question);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        // QuestionPager
        qp = findViewById(R.id.question_pager);
        qa = new QuestionAdapter(this);
        qp.setAdapter(qa);
        qp.setCurrentItem(0);
        qp.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);

                if (Coordinator.inProcess && position == qa.getItemCount() - 1) {
                    Coordinator.check();
                    Objects.requireNonNull(getSupportActionBar()).setTitle(R.string.finish_label);
                } else {
                    Objects.requireNonNull(getSupportActionBar()).setTitle(
                        MessageFormat.format(
                            getResources().getString(R.string.question_label), position + 1
                        )
                    );
                }
            }
        });
        qp.setPageTransformer(new MyPageTransformer());
        //
    }

    @Override
    protected void onResume() {
        super.onResume();
        // If theme changed -> apply new theme
        if (theme != new LOSTheme(this).current) recreate();
        //
    }

    @Override
    public void finish() {
        Coordinator.resetScore();
        startActivity(new Intent(this, MainActivity.class));
    }

    @Override
    public void next(boolean isCorrect) {
        Coordinator.calcScore(isCorrect);
        Toast.makeText(getApplicationContext(), (isCorrect) ? R.string.question_right :
            R.string.question_wrong, Toast.LENGTH_SHORT).show();
        qp.setCurrentItem(qp.getCurrentItem() + 1, true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.los_menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            startActivity(new Intent(this, LOSSettingsActivity.class));
        }
        return super.onOptionsItemSelected(item);
    }

    public void onBackPressed() {
        if ((qp.getCurrentItem() > 0)) qp.setCurrentItem(qp.getCurrentItem() - 1, true);
        else startActivity(new Intent(this, MainActivity.class));
    }
}

File: app/src/main/java/loshica/quiz/QuestionActivity.kt
package loshica.quiz

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import loshica.quiz.databinding.ActivityQuestionBinding
import loshica.quiz.interfaces.FinishFragmentHandler
import loshica.quiz.interfaces.QuestionFragmentHandler
import loshica.quiz.view.MyPageTransformer
import loshica.quiz.view.QuestionAdapter
import loshica.quiz.viewModel.*
import loshica.vendor.LOSActivity
import java.text.MessageFormat

class QuestionActivity : LOSActivity(), QuestionFragmentHandler, FinishFragmentHandler {

    private val nameArg: String = "name"
    private lateinit var qa: QuestionAdapter
    private lateinit var b: ActivityQuestionBinding

    private lateinit var game: GameModel
    private lateinit var gameFactory: GameModelFactory
    private val question: QuestionModel by viewModels()
    private val storage: StorageModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        b = ActivityQuestionBinding.inflate(layoutInflater)
        gameFactory = GameModelFactory(intent.getStringExtra(nameArg)!!.toString(), application)
        game = ViewModelProvider(this, gameFactory).get(GameModel::class.java)

        setContentView(b.root)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)!!

        // QuestionPager
        qa = QuestionAdapter(question.getQuestions(), this)
        b.qPager.adapter = qa
        b.qPager.currentItem = 0
        b.qPager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)

                if (game.inProcess() && position == qa.itemCount - 1) {
                    storage.check(game.getPlayer())
                    game.cancel()
                    supportActionBar?.setTitle(R.string.finish_label)!!
                } else {
                    supportActionBar?.title = MessageFormat.format(
                        resources.getString(R.string.question_label), position + 1
                    )
                    if (question.isLeftSwipe(position)) question.setCounter(position)
                }
            }
        })
        b.qPager.setPageTransformer(MyPageTransformer())
        //
    }

    override fun finish() { startActivity(Intent(this, MainActivity::class.java)) }

    override fun next(isCorrect: Boolean) {
        game.calcScore(isCorrect)
        question.setCounter(b.qPager.currentItem + 1)
        Toast.makeText(applicationContext, question.toastText(b.qPager.currentItem), Toast.LENGTH_SHORT)
            .show()
        b.qPager.setCurrentItem(b.qPager.currentItem + 1, true)
    }

    override fun onBackPressed() {
        if (b.qPager.currentItem > 0) b.qPager.setCurrentItem(b.qPager.currentItem - 1, true)
        else startActivity(Intent(this, MainActivity::class.java))
    }
}